content/en/blog/_posts/TBD-admission-controllers-for-container-drift/index.md

```diff
@@ -0,0 +1,90 @@
+
+## A new kubectl plugin for better user experience
+Our admission controller service works great for solving the container drift issue we had on the platform. It is able able to submit all related K8s events to the target pod that has been affected. However, as K8s events often get away quickly (one hour of persistency by the default configuration), we need to provide other ways for developers to get their pod interaction activity. A [kubectl plugin](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) is a perfect choice for us to expose this information. We name our plugin “kubectl-pi” (short for pod-interaction) and provide two subcommands: “get” and “extend”. When the “get” command is called, the plugin simply checks the metadata attached by our admission controller service and transfers it to human-readable information. Here is an example output of the `kubectl pi get` command:
+
+```
+$ kubectl pi get test-pod
+POD_NAME  INTERACTOR  POD_TTL  EXTENSION  EXTENSION_REQUESTER  EVICTION_TIME
+test-pod  username-1  1h0m0s   /          /                    2021-09-24 17:00:00 -0800 PST
+```
+
+The plugin can also be used to extend the current eviction time of a pod, in case developers need more buffer to debug ongoing issues after executing on their containers. This is achieved by the `kubectl pi extend` command, where the plugin patches relevant K8s annotation to the given pod. This annotation includes the duration and username who made the extension request for transparency (also displayed in the table returned from the “get” command). There is another webhook defined in our admission controller service that admits annotation updates and resets the eviction timer of the target pod accordingly. An example of requesting the extension from the developer side would be:
+
+```
+$ kubectl pi extend test-pod --duration=30m
+Successfully extended the termination time of pod/test-pod with a duration=30m
+ 
+$ kubectl pi get test-pod
+POD_NAME  INTERACTOR  POD_TTL  EXTENSION  EXTENSION_REQUESTER  EVICTION_TIME
+test-pod  username-1  1h0m0s   30m        username-2           2021-09-24 17:30:00 -0800 PST
+```
+
+## Future improvement
+Although our admission controller service works great in handling interactive requests to a pod, it could as well evict the pod while the actual commands are no-op in these requests. For instance, developers sometimes run `kubectl exec` merely to check their service logs stored on hosts. Nevertheless, the target pods would still get bounced despite the state of their containers not changing at all. One of the improvements here could be adding the ability to distinguish the commands that are passed to the interactive requests, so that no-op commands should not always force a pod eviction. However, this becomes challenging when developers get a shell to a running container and execute commands inside the shell, since they will not be visible to our admission controller service anymore.
```

Comment:
This would be a good place to mention areas where the existing code deviates from good practice for extending Kubernetes (table headings in kubectl output, annotations vs. labels if the app code itself gets this wrong, anything else).

With a brief writeup of the gaps here, and fixes elsewhere, I think this'd be good to merge. To be clear, you don't need to fix kube-exec-controller - you just need to point out what you'd do differently if you were writing code that would become part of Kubernetes.

I think mentioning these gaps is important as we wouldn't want to publish this onto our official blog without pointing that out.

Reply:
@sftim Thanks for pointing this out.
For the "table headings in kubectl output," I have updated the example output in this blog post and also submitted a [PR](https://github.com/box/kube-exec-controller/pull/3) in the plugin source code repo to fix it. I believe we can get the PR merged before publishing this blog post (cc @FraBle).

For "annotations vs. labels," it was actually part of the initial design, and I have replied in one of your above comments with more detail. Please let me know if that makes sense to you. If so, I will leave this "future improvement" section as is. Thanks.

Reply:
Thanks for the heading fix for the plugin.

On labels vs annotations I get why you'd want there to be _a_ label (for example, named `box.com/pod-time-to-live`), but not why (for example) `box.com/podInitialInteractionTimestamp` is a label and not an annotation. I can't see you querying for the value it has, and this would only ever be set in conjunction with the TTL label.

Reply:
@sftim No problem, glad to improve the plugin code to be more kubectl native.

I think what you said above on labels vs annotations makes sense, as we only used the `box.com/podInitialInteractionTimestamp` label to identify the affected pods but not actually querying for its value. 

I have updated this "Future improvement" section by mentioning the existing gaps to the ideal K8s design (in [this](https://github.com/kubernetes/website/pull/29761/commits/540390b811ac7f193b8cc62d71f4647db8999dd0) latest commit).